### PR TITLE
Make autopacking faster

### PR DESCRIPTION
**Pull request by b-chu**
Adds a profiling flag to collators to avoid unnecessary padding.

1. Seq2Seq collator pads to the length of the longest processed example instead of max_seq_len
2. The packing collator doesn't concatenate tensors when profiling
3. The packing collator returns early instead of repadding the freshly concatenated tensors to max_seq_len

Tested with reference runs:

`autopacking-oom-mpt-baseline-Pr7MgS`
- global_train_batch_size: 256
- device_train_microbatch_size: 16
- num_gpus: 8
- max_seq_len: 131072
- packing ratio: 1310.7
- PR packing ratio: 1310.7

`autopacking-oom-mpt-baseline-CPadi0`
- global_train_batch_size: 8
- device_train_microbatch_size: 1
- num_gpus: 8
- max_seq_len: 32768
- packing ratio: 190.1
- PR packing ratio: 190.1

- global_train_batch_size: 64
- device_train_microbatch_size: 1
- num_gpus: 64
- max_seq_len: 131072
- PR packing ratio: 897.1 `autopacking-oom-mpt-fix-mcV61Z`

### CODE REVIEW

File: llmfoundry/data/finetuning/collator.py [hunk 1]
```diff
@@ -320,24 +323,29 @@ def _process_and_batch_decoder_only(
...
-            )
+        input_ids_and_labels = [stitch_turns_decoder_only(
+            example_turns=example['turns'],
+            target_prompts=self.target_prompts,
+            target_responses=self.target_responses,
+            eos_token_id=self.tokenizer.eos_token_id,
+        ) for example in examples]
+
+        if self._is_profiling:
+            max_seq_len = max([len(input_ids) for input_ids, _ in input_ids_and_labels])
```
Comment (irenedea) [inline llmfoundry/data/finetuning/collator.py:334]: Is this doing almost the same thing as `allow_pad_trimming`? @alextrott16, do you have any thoughts here on how/if we can consolidate this functionality? We're avoiding padding to the max seq length because it's too slow at super long seq length and with packing, we trim the padding off anyway.

File: llmfoundry/data/finetuning/collator.py [hunk 2]
```diff
@@ -235,6 +235,7 @@ def __init__(
         target_prompts: str = 'none',
         allow_pad_trimming: bool = False,
         batch_metadata: Optional[Dict[str, Any]] = None,
+        is_profiling: bool = False,
```
Comment (irenedea) [inline llmfoundry/data/finetuning/collator.py:238]: let's rename this flag. Would prefer a flag that describes the functionality of what happens rather than an overall state so that the flag makes sense if you use the collator independently. 

Ideally, if the below comment makes sense, we can just use the allow_pad_trimming flag

File: llmfoundry/data/packing.py [hunk 3]
```diff
@@ -277,43 +309,21 @@ def _first_fit_bin_packing(
...
-    keys = packed_examples[0].keys()
-    batch = {}
-    for key in keys:
-        batch[key] = torch.stack([
-            pad_tensor(example[key], pad_vals[key])
-            for example in packed_examples
-        ])
-    return batch
+def _pad_tensor(tensor: torch.Tensor, pad_value: int, max_seq_len: int, padding_side: str):
+    return tensor
```
Comment (irenedea) [inline llmfoundry/data/packing.py:313]: remove?

File: llmfoundry/data/packing.py [hunk 4]
```diff
@@ -367,6 +379,8 @@ def auto_packing_ratio(
         num_packing_ratios=num_packing_ratios,
         device_batch_size=device_batch_size,
     )
+    dist.barrier()
```
Comment (irenedea) [inline llmfoundry/data/packing.py:382]: are you planning to keep this dist barrier?

File: llmfoundry/data/packing.py [hunk 5]
```diff
@@ -30,6 +30,7 @@ def __init__(
         pad_token_id: int,
         padding_side: Literal['left', 'right'],
         max_leftover_bins_to_keep: Optional[int] = None,
+        is_profiling: bool = False,
```
Comment (irenedea) [inline llmfoundry/data/packing.py:33]: can you add a doc string for this class and its args? ty!

File: llmfoundry/data/packing.py [hunk 6]
```diff
@@ -183,6 +211,7 @@ def _first_fit_bin_packing(
     num_bins: int,
     max_bin_size: int,
     existing_bins: List[Tuple[int, Dict[str, torch.Tensor]]],
+    is_profiling: bool = False,
```
Comment (irenedea) [inline llmfoundry/data/packing.py:214]: I think it'll make more sense to have this as a class method and just use `self.is_profiling`

Comment (irenedea) [reply llmfoundry/data/packing.py:214]: Similar reason to wanting to rename the is_profiling param in Seq2SeqCollator. It doesn't make a lot of sense if you look at this function individually.

File: llmfoundry/data/finetuning/collator.py (hunk 2 above)
Comment (irenedea) [reply llmfoundry/data/finetuning/collator.py:238]: also since this is only supported for decoder_only_format, please add a checks that errors if this is not the case

File: llmfoundry/data/finetuning/collator.py (hunk 1 above)
Comment (alextrott16) [reply llmfoundry/data/finetuning/collator.py:334]: As discussed offline, yeah this has _nearly_ the same behavior as `allow_pad_trimming` but the latter doesn't trim the first batch. And that's a problem during profiling since it just grabs a single batch.

So, I recommend sticking with the way @b-chu has it implemented here :)

File: llmfoundry/data/packing.py (hunk 6 above)
Comment (b-chu) [reply llmfoundry/data/packing.py:214]: Done

File: llmfoundry/data/packing.py (hunk 4 above)
Comment (b-chu) [reply llmfoundry/data/packing.py:382]: Removed

File: llmfoundry/data/packing.py (hunk 3 above)
Comment (b-chu) [reply llmfoundry/data/packing.py:313]: Removed

File: llmfoundry/data/packing.py (hunk 5 above)
Comment (b-chu) [reply llmfoundry/data/packing.py:33]: Done

File: llmfoundry/data/finetuning/collator.py (hunk 2 above)
Comment (b-chu) [reply llmfoundry/data/finetuning/collator.py]: Renamed and added support for _process_and_batch_encoder_decoder

File: llmfoundry/data/finetuning/collator.py [hunk 7]
```diff
@@ -320,24 +327,33 @@ def _process_and_batch_decoder_only(
...
                 example_turns=example['turns'],
                 target_prompts=self.target_prompts,
                 target_responses=self.target_responses,
                 eos_token_id=self.tokenizer.eos_token_id,
-            )
+            ) for example in examples
+        ]
+
+        if self._pad_to_longest:
+            max_seq_len = max([
```
Comment (irenedea) [inline llmfoundry/data/finetuning/collator.py:340]: You want to also include the true max seq len here
```suggestion
            max_seq_len = max([self.max_seq_len] + [
```

File: llmfoundry/data/packing.py [hunk 8]
```diff
@@ -492,6 +510,7 @@ def profile(raw_batch_size: int) -> Tuple[Optional[float], Optional[float]]:
             pad_token_id=0,  # <-- Doesn't need to be correct for profiling
             padding_side='left',  # <-- Doesn't need to be correct for profiling
             max_leftover_bins_to_keep=max_leftovers_to_keep,
+            is_profiling=False,
```
Comment (irenedea) [inline llmfoundry/data/packing.py:513]: shouldn't this be true?

Comment (b-chu) [reply llmfoundry/data/packing.py:513]: Done

File: llmfoundry/data/finetuning/collator.py (hunk 7 above)
Comment (irenedea) [reply llmfoundry/data/finetuning/collator.py:340]: oh my bad it should be 
```
max_seq_len = min(self.max_seq_len, max([...]))
```
right?

Comment (irenedea) [reply llmfoundry/data/finetuning/collator.py:340]: I think the test was failing because the examples were not getting truncated

File: llmfoundry/data/packing.py [hunk 9]
```diff
@@ -177,143 +317,22 @@ def _combine_in_place(
...
-    batch = {}
-    for key in keys:
-        batch[key] = torch.stack([
-            pad_tensor(example[key], pad_vals[key])
-            for example in packed_examples
-        ])
-    return batch
+def _pad_tensor(
+    tensor: torch.Tensor, pad_value: int, max_seq_len: int, padding_side: str
+):
```
[the remark this answers is not in the document]
Comment (b-chu) [reply llmfoundry/data/packing.py:328]: Done

File: llmfoundry/data/packing.py [hunk 10]
```diff
@@ -20,7 +20,17 @@
...
+    """Utility collator for packing to reduce padding.
+
+    Args:
+        collator (Callable): The base collator to use.
+        target_batch_size (int): The number of bins.
+        max_seq_len(int): The maximum sequence length of a bin.
+        pad_token_id (int): The padding token id.
+        padding_side (Literal['left', 'right']): The side to pad on.
+        max_leftover_bins_to_keep (Optional[int]): The number of leftover bins to keep.
+        is_profiling (bool): Whether the collator is being used for profiling.
```
[the remark this answers is not in the document]
Comment (b-chu) [reply llmfoundry/data/packing.py:32]: Done

File: llmfoundry/data/packing.py [hunk 11]
```diff
@@ -115,15 +130,140 @@ def _pack_trimmed_examples(
         self.n_packed_examples += self.out_size
         self._leftover_bins = leftover_bins[:self.max_leftover_bins_to_keep]
 
+        if self._is_profiling:
```
[the remark this answers is not in the document]
Comment (b-chu) [reply llmfoundry/data/packing.py:133]: Done

File: llmfoundry/data/finetuning/collator.py (hunk 7 above)
Comment (b-chu) [reply llmfoundry/data/finetuning/collator.py:340]: Yea fixed it

**Review by irenedea — APPROVED** [review]
lgtm! just remove print statements :)

File: llmfoundry/data/packing.py [hunk 12]
```diff
@@ -115,15 +130,140 @@ def _pack_trimmed_examples(
...
+                bins.append((size, example))
+                continue
+
+            # Add it to the first bin it fits in
+            added = False
+            for bidx in range(len(bins)):
+                if bins[bidx][0] + size <= max_bin_size:
+                    bin_size, packed_example = bins.pop(bidx)
+                    bin_size = bin_size + size
+                    if not self._is_profiling:
```
[the remark this answers is not in the document]
Comment (irenedea) [reply llmfoundry/data/packing.py:239]: discussed offline, just add some comments describing that we aren't going to combine the tensor when profiling for efficiency so we just return the first example, which isn't "correct"

File: llmfoundry/data/packing.py [hunk 13]
```diff
@@ -86,24 +102,28 @@ def pack(self, batch: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
             ]
         # Cut everything down to size
         sizes, trimmed_examples = _trim_batch(batch)
-        return self._pack_trimmed_examples(trimmed_examples, sizes)
+        packed_batch = self._pack_trimmed_examples(trimmed_examples, sizes)
+        assert packed_batch is not None
```
Comment (irenedea) [inline llmfoundry/data/packing.py:106]: use ValueError over assert
```suggestion
        if packed_batch is None:
            raise ValueError(...)
```

Comment (b-chu) [reply llmfoundry/data/packing.py:106]: This is purely for pyright, not actual error handling

Comment (irenedea) [reply llmfoundry/data/packing.py:106]: also can you add 
```
if self._is_profiling:
    raise ValueError('`pack` should not be called while profiling. Please use `_pack_trimmed_examples`.')
```

Comment (b-chu) [reply llmfoundry/data/packing.py:106]: `pack` is never called when profiling

Comment (irenedea) [reply llmfoundry/data/packing.py:106]: So that someone doesn't call `pack` when creating a BinPackCollator in profiling mode
